pages/blog/unleash-the-power-of-fragments-with-graphql-codegen.mdx

```diff
@@ -0,0 +1,314 @@
+import withArticle from '../../ui/blog/article'
+
+export const meta = {
+  title: 'Unleash the power of Fragments with GraphQL Codegen',
+  author: 'laurin',
+  tags: [
+    'graphql',
+    'codegen',
+    'react',
+    'typescript',
+    'relay',
+    'apollo',
+    'urql'
+  ],
+  date: '2022-08-03',
+  description:
+    "The most important parts of Relay are the concepts of building and scaling applications, let's show how you can use these patterns in your existing project, no matter what client library you are currently using.",
+  image:
+    '/blog-assets/unleash-the-power-of-fragments-with-graphql-codegen/cover.png',
+  thumbnail:
+    '/blog-assets/unleash-the-power-of-fragments-with-graphql-codegen/thumbnail.png'
+}
+
+export default withArticle(meta)
+
+There is no doubt that Relay is the most advanced JavaScript GraphQL Client that is out there. However, at the same time, the learning curve and adoption is yet very sparse compared to other popular alternatives such as Apollo Client, Urql, or GraphQL Request.
+
+As a consequence, many people don't even know about all the benefits and patterns used within (and so far being exclusive to) Relay.
+The Guild gathered and worked on projects whose GraphQL client usage spans across all the available solutions today.
+In our opinion, the most important parts of Relay are the concepts of building and scaling applications.
+These patterns are actually applicable to any GraphQL Client, you just need the right code generation tool for the job.
+
+On a high level these benefits are the following:
+
+**A fragmentized component tree.** Colocating the GraphQL operation and fragment definitions with the components and composing them up for only sending a single request to our server.
+
+![Fragmentized Component Tree](/blog-assets/unleash-the-power-of-fragments-with-graphql-codegen/component-tree-diagram.png)
+
+**Colocate GraphQL documents with the component code.** Instead of writing GraphQL operations in a dedicated file, the operations are written within the component code.
+
+```tsx
+import { gql, useFragment, FragmentType } from './gql'
+
+const Avatar_UserFragment = gql(/* GraphQL */ `
+  fragment Avatar_UserFragment on User {
+    avatarUrl
+  }
+`)
+
+type AvatarProps = {
+  user: FragmentType<typeof Avatar_UserFragment>
+}
+
+export function Avatar(props: AvatarProps) {
+  const user = useFragment(Avatar_UserFragment, props.user)
+  return <CircleImage src={user.avatarUrl} />
+}
+```
+
+**Data (fragment) masking.** Ensure that a componenent can only access the data defined in it's fragment definitions in order to keep the component a self-contained building block that can be re-used.
+
+---
+
+We are happy to announce that we now have a preset for bringing the above Relay patterns to your existing projects, no matter what client library you are currently using, without having to switch over and commit to the Relay eco-system.
+
```

Comment:
Can we also mention that this is now our recommended way for using GraphQL and Typescript on the frontend?

Reply:
done